pkg/engine/engine.go

```diff
@@ -492,10 +584,135 @@ func (e *Engine) detectorWorker(ctx context.Context) {
+// by the same detector in the chunk. Exact matches on lookup indicate a duplicate secret for a detector
+// in that chunk - which is expected and not malicious. Those intra-detector dupes are still verified.
+type chunkSecretKey struct {
+	secret       string
+	detectorInfo ahocorasick.DetectorInfo
+}
+
+func likelyDuplicate(ctx context.Context, val chunkSecretKey, dupes map[chunkSecretKey]struct{}) bool {
+	const similarityThreshold = 0.9
+
+	valStr := val.secret
+	for dupeKey := range dupes {
+		dupe := dupeKey.secret
+		// Avoid comparing strings of vastly different lengths.
+		if len(dupe)*10 < len(valStr)*9 || len(dupe)*10 > len(valStr)*11 {
+			continue
+		}
+
+		similarity := strutil.Similarity(valStr, dupe, metrics.NewLevenshtein())
+
+		// close enough
+		if similarity > similarityThreshold {
+			ctx.Logger().V(2).Info(
+				"found similar duplicate",
+				"val", val,
+				"dupe", dupe,
+				"similarity", similarity,
+			)
+			return true
+		}
+	}
+	return false
+}
+
+func (e *Engine) verificationOverlapWorker(ctx context.Context) {
+	var wgDetect sync.WaitGroup
+
+	// Reuse the same map and slice to avoid allocations.
+	const avgSecretsPerDetector = 8
+	detectorsWithResult := make(map[detectors.Detector]struct{}, avgSecretsPerDetector)
+	chunkSecrets := make(map[chunkSecretKey]struct{}, avgSecretsPerDetector)
+
+	for chunk := range e.verificationOverlapChunksChan {
+		for _, detector := range chunk.detectors {
+			// DO NOT VERIFY at this stage of the pipeline.
+			results, err := detector.FromData(ctx, false, chunk.chunk.Data)
+			if err != nil {
+				ctx.Logger().Error(err, "error verifying chunk")
+			}
+
+			if len(results) == 0 {
+				continue
+			}
+			if _, ok := detectorsWithResult[detector]; !ok {
+				detectorsWithResult[detector] = struct{}{}
+			}
+
+			for _, res := range results {
+				var val []byte
+				if res.RawV2 != nil {
+					val = res.RawV2
+				} else {
+					val = res.Raw
+				}
+
+				// Use levenstein distance to determine if the secret is likely the same.
+				// Ex:
+				// - postman api key: PMAK-qnwfsLyRSyfCwfpHaQP1UzDhrgpWvHjbYzjpRCMshjt417zWcrzyHUArs7r
+				// - malicious detector "api key": qnwfsLyRSyfCwfpHaQP1UzDhrgpWvHjbYzjpRCMshjt417zWcrzyHUArs7r
+				key := chunkSecretKey{secret: string(val), detectorInfo: detector}
+				if _, ok := chunkSecrets[key]; ok {
+					chunkSecrets[key] = struct{}{}
+					continue
+				}
+
+				if likelyDuplicate(ctx, key, chunkSecrets) {
+					// This indicates that the same secret was found by multiple detectors.
+					// We should NOT VERIFY this chunk's data.
+					if e.verificationOverlapTracker != nil {
+						e.verificationOverlapTracker.increment()
+					}
+					e.processResult(ctx, detectableChunk{
+						chunk:    chunk.chunk,
+						detector: detector,
+						decoder:  chunk.decoder,
+						wgDoneFn: wgDetect.Done,
+					}, res)
+
+					// Remove the detector from the list of detectors with results.
+					delete(detectorsWithResult, detector)
```

Comment:
And won't this only remove later detectors in a collision set, but not the first? (Assuming my previous concern about `likelyDuplicate` always returning true is misplaced, anyway.)

Reply:
Yep, you are correct. The `detectorsWithResult` should be a `map[ahocorasick.DetectorInfo]struct{}` instead.